src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionHelper.cs

```diff
@@ -37,86 +38,158 @@ public static string GetExpressionText(LambdaExpression expression, ExpressionTe
-                {
-                    var memberExpressionPart = (MemberExpression)part;
-                    var name = memberExpressionPart.Member.Name;
-
-                    // If identifier contains "__", it is "reserved for use by the implementation" and likely compiler-
-                    // or Razor-generated e.g. the name of a field in a delegate's generated class.
-                    if (name.Contains("__"))
-                    {
-                        // Exit loop. Should have the entire name because previous MemberAccess has same name as the
-                        // leftmost expression node (a variable).
+                        // If identifier contains "__", it is "reserved for use by the implementation" and likely
+                        // compiler- or Razor-generated e.g. the name of a field in a delegate's generated class.
+                        if (name.Contains("__"))
+                        {
+                            // Exit loop.
+                            part = null;
+                        }
+                        else
+                        {
+                            lastIsModel = string.Equals("model", name, StringComparison.OrdinalIgnoreCase);
+                            length += name.Length + 1;
+                            part = memberExpressionPart.Expression;
+                            segmentCount += 2;
+                            trailingMemberExpressions++;
+                        }
                         break;
-                    }
 
-                    builder.Insert(0, name);
-                    builder.Insert(0, '.');
-                    part = memberExpressionPart.Expression;
-                }
-                else
-                {
-                    break;
+                    case ExpressionType.Parameter:
+                        // Unsupported but indicates previous member access was not the view's Model.
+                        lastIsModel = false;
+                        part = null;
+                        break;
+
+                    default:
+                        // Unsupported.
+                        part = null;
+                        break;
                 }
             }
 
-            // If parts start with "model", then strip that part away.
-            if (part == null || part.NodeType != ExpressionType.Parameter)
+            // Expression must contain indexers if not all parts are member expressions.
+            var containsIndexers = (trailingMemberExpressions * 2) != segmentCount;
+
+            // If name would start with ".model", then strip that part away.
+            if (lastIsModel)
+            {
+                length -= 6;    // ".model".Length
+                segmentCount -= 2;
+                trailingMemberExpressions--;
+            }
+
+            // Trim the leading "." if present.
+            if (trailingMemberExpressions > 0)
+            {
+                length--;
+                segmentCount--;
```

Comment:
`'.'` is no longer a segment. `segmentCount` now matches iteration count of second loop and has a few associated comments.